File: phone-app/android/app/src/main/java/com/example/phone_app/background/BackgroundDecryptor.java
package com.example.phone_app.background;

import android.util.Log;

import com.example.phone_app.CiphertextWrapper;
import com.example.phone_app.JNIBridge;
import com.example.phone_app.background.serialization.ContactMessage;
import com.example.phone_app.background.serialization.NewDistanceMessage;
import com.example.phone_app.background.serialization.NewPartialMessage;
import com.example.phone_app.config.Config;
import com.example.phone_app.config.EncryptionType;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class BackgroundDecryptor {
  private final static ExecutorService decryptorService = Executors.newFixedThreadPool(5);
  private final static JNIBridge bridge = new JNIBridge();

  public static void getDistances(String userId, char[] privateKey, boolean partial) {
    while (true) {
      List<NewDistanceMessage> givenCiphertexts = ConnectionService.getDistances(userId, partial);
      if (givenCiphertexts.size() == 0) {
        return;
      }
      Log.d(BackgroundDecryptor.class.getName(), "I have received " + givenCiphertexts.size() + " ciphertexts.");
      if (partial) {
        computePartial(givenCiphertexts, privateKey);
      } else {
        wasInContact(givenCiphertexts, privateKey, userId);
      }
    }
  }

  public static void computePartial(List<NewDistanceMessage> givenCiphertexts, char[] privateKey) {
    for (NewDistanceMessage distanceMessage : givenCiphertexts) {
      decryptorService.execute(() -> {
        Log.d(BackgroundDecryptor.class.getName(), "Start timestamp " + distanceMessage.getTimestamp());
        Log.d(BackgroundDecryptor.class.getName(), "End timestamp " + distanceMessage.getTimestampEnd());
        Log.d(BackgroundDecryptor.class.getName(), "Row id " + distanceMessage.getRowId());
        String partialDistanceString = ((CiphertextWrapper) bridge.decryptMulti(distanceMessage.getCiphertext().toCharArray(), "".toCharArray(), privateKey, false)).getLatitudeCos();
        String partialAltitudeDifference = ((CiphertextWrapper) bridge.decryptMulti(distanceMessage.getAltitudeDifference().toCharArray(), "".toCharArray(), privateKey, false)).getLatitudeCos();
        ConnectionService.sendObject(new NewPartialMessage(partialDistanceString, partialAltitudeDifference, distanceMessage.getRowId()), "new-partial-distance");
      });
    }
  }

  private static void wasInContact(List<NewDistanceMessage> givenCiphertexts, char[] privateKey, String userId) {
    for (NewDistanceMessage distanceMessage : givenCiphertexts) {
      decryptorService.execute(() -> {
        Log.d(BackgroundDecryptor.class.getName(), "Start timestamp " + distanceMessage.getTimestamp());
        Log.d(BackgroundDecryptor.class.getName(), "End timestamp " + distanceMessage.getTimestampEnd());
        Log.d(BackgroundDecryptor.class.getName(), "Contact user id " + distanceMessage.getContactUserId());
        Log.d(BackgroundDecryptor.class.getName(), "Current ciphertext size is " + distanceMessage.getCiphertext().length());

        double initialResult;
        double differenceInAltitude;

        if (Config.getEncryptionType() == EncryptionType.LATTIGO_MK) {
          initialResult = (Double) bridge.decryptMulti(distanceMessage.getCiphertext().toCharArray(), distanceMessage.getPartialDistance().toCharArray(),
                  privateKey, true);
          differenceInAltitude = (Double) bridge.decryptMulti(distanceMessage.getAltitudeDifference().toCharArray(), distanceMessage.getPartialAltitudeDifference().toCharArray(),
                  privateKey, true);
        } else {
          initialResult = bridge.decrypt(distanceMessage.getCiphertext().toCharArray(), privateKey);
          differenceInAltitude = bridge.decrypt(distanceMessage.getAltitudeDifference().toCharArray(), privateKey);
        }
        Log.d(BackgroundDecryptor.class.getName(), "Altitude difference was " + differenceInAltitude);
        if (initialResult < 0) {
          Log.d(BackgroundDecryptor.class.getName(), "The location was very close");
        }
        initialResult = Math.asin(Math.sqrt(initialResult / 2.0)) * 6378.8 * 2.0 * 1000;
        Log.d(BackgroundDecryptor.class.getName(), "The distance was " + initialResult + " meters.");
        if ((initialResult <= 6 || Double.isNaN(initialResult)) && differenceInAltitude * 100 < 210) {
          Log.d(BackgroundDecryptor.class.getName(), "Found contact " + distanceMessage.getContactUserId());
          ConnectionService.sendObject(new ContactMessage(userId, distanceMessage.getContactUserId(),
                  distanceMessage.getTimestamp(), distanceMessage.getTimestampEnd()), "new-contact");
        }
      });
    }
  }
}

File: phone-app/ios/Runner/background/BackgroundDecryptor.swift
//
//  BackgroundDecryptor.swift
//  Runner
//
//  Created by Andru Stefanescu on 14.12.2021.
//

import Foundation

class BackgroundDecryptor {
    private static let bridge: HeBridgeWrapper = HeBridgeWrapper()
    
    public static func getDistances(userId: String, privateKey: String, partial: Bool) {
        DispatchQueue.global(qos: .background).async {
            while(true) {
                let givenCiphertexts : Array<NewDistanceMessage> = ConnectionService.getDistances(userId: userId, partial: partial)
                if (givenCiphertexts.isEmpty) {
                    return
                }
                NSLog("I have received %d ciphertexts", givenCiphertexts.count)
                if (partial) {
                    computePartial(givenCiphertexts: givenCiphertexts, privateKey: privateKey, userId: userId)
                } else {
                    wasInContact(givenCiphertexts: givenCiphertexts, privateKey: privateKey, userId: userId)
                }
            }
        }
    }
    
    private static func wasInContact(givenCiphertexts: Array<NewDistanceMessage>, privateKey: String, userId: String) {
        for distanceMessage in givenCiphertexts {
            NSLog("Start timestamp \(distanceMessage.timestamp)")
            NSLog("End timestamp \(distanceMessage.timestampEnd)")
            NSLog("Contact user id \(distanceMessage.contactUserId)")
            NSLog("Current ciphertext size is \(distanceMessage.ciphertext.count)")
            
            var initialResult: Double
            var altitudeDifference: Double
            
            if (Config.getEncryptionType() == EncryptionType.LATTIGO_MK) {
                initialResult = Double(truncating: bridge.decryptMulti(distanceMessage.ciphertext,partialCipher: distanceMessage.partialDistance, privateKey: privateKey, isFinal: true) as! NSNumber)
                altitudeDifference = Double(truncating: bridge.decryptMulti(distanceMessage.altitudeDifference,partialCipher: distanceMessage.partialAltitudeDifference, privateKey: privateKey, isFinal: true) as! NSNumber)
            } else {
                initialResult = bridge.decrypt(distanceMessage.ciphertext, privateKey: privateKey)
                altitudeDifference = bridge.decrypt(distanceMessage.altitudeDifference, privateKey: privateKey)
            }
            
            NSLog("Altitude difference was \(altitudeDifference)")
            initialResult = asin(sqrt(initialResult / 2.0)) * 6378.8 * 2.0 * 1000
            
            if (initialResult.isNaN) {
                initialResult = 0
            }
            
            NSLog("Distance was \(initialResult)")
            
            if (initialResult <= 6.0 && altitudeDifference * 100 < 210) {
                NSLog("Found contact \(distanceMessage.contactUserId)")
                ConnectionService.sendNewContact(contact: ContactMessage(userId: userId, infectedUserId: distanceMessage.contactUserId, timestamp: distanceMessage.timestamp, timestampEnd: distanceMessage.timestampEnd))
            }
        }
    }
    
    public static func computePartial(givenCiphertexts: Array<NewDistanceMessage>, privateKey: String, userId: String) {
        for distanceMessage in givenCiphertexts {
            NSLog("Start timestamp \(distanceMessage.timestamp)")
            NSLog("End timestamp \(distanceMessage.timestampEnd)")
            NSLog("Contact user id \(distanceMessage.contactUserId)")
            NSLog("Current ciphertext size is \(distanceMessage.ciphertext.count)")
            let partialStringDistance: String = bridge.decryptMulti(distanceMessage.ciphertext, partialCipher: "", privateKey: privateKey, isFinal: false) as! String
            let partialStringAltitudeDifference: String = bridge.decryptMulti(distanceMessage.altitudeDifference, partialCipher: "", privateKey: privateKey, isFinal: false) as! String
            ConnectionService.sendNewPartial(partialMessage: NewPartialMessage(partialDistance: partialStringDistance, partialAltitudeDifference: partialStringAltitudeDifference, rowId: distanceMessage.rowId))
        }
    }
}
